internal/logger.go

```diff
@@ -15,37 +15,134 @@
-// everything fails, it will just output to the standard error channel.
-func GetLoggerFile() *os.File {
-	// Determine the path to be used.
-	var path string
-	if env := os.Getenv(logEnv); env != "" {
-		path = env
-	} else {
-		path = defaultLogPath
+// Default log file.
+const DefaultLogFile = "suseconnect.log"
+
+// Default path for the log file.
+const DefaultLogPath = "/var/log/suseconnect.log"
+
+// Environment variable used to specify a custom path for the log file.
+const LogEnv = "SUSECONNECT_LOG_FILE"
+
+// getLogEnv return the value set to the [LogEnv] environment variable.
+func getLogEnv() string {
+	return strings.TrimSpace(os.Getenv(LogEnv))
+}
+
+// isWritable checks if the path can be open and written to.
+// If the file does not exists it attemps to create.
+func isWritable(path string) bool {
+	lf, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0640)
+
+	if err != nil {
+		return false
 	}
 
-	// If it's writable, use the given file, otherwise use os.Stderr.
-	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0640)
+	lf.Close()
+
+	return true
+}
+
+// getValidLogPath returns a writable location for the log file.
+//
+// If the path does not exists it attempst to create. Failing to
+// create a writable file, it returns [DefaultLogPath].
+// If the path is a directory a filename is appended to it. Failing
+// to create the directory, it returns [DefaultLogPath].
+// If the path is not a regulat file (e.g device file) it
+// returns [DefaultLogPath].
+func getValidLogPath(path string) string {
+	path = strings.TrimSpace(path)
+
+	if len(path) == 0 {
+		return DefaultLogPath
+	}
+
+	if !filepath.IsAbs(path) {
+		absPath, err := filepath.Abs(path)
+
+		if err != nil {
+			log.Printf("Failed to get absolute path for '%s': %v\n", path, err)
+			return DefaultLogPath
+		}
+
+		path = absPath
+	}
+
+	// if path is a valid directory, we must append a file name
+	// to it because filepath.Split is unaware and will potentially
+	// assign the dirname to filename
+	if fi, err := os.Stat(path); err == nil {
+		if fi.Mode().IsDir() {
+			path = filepath.Join(path, DefaultLogFile)
+		}
+	}
+
+	dir, filename := filepath.Split(path)
+
+	if len(filename) == 0 {
+		filename = DefaultLogFile
+	}
+
+	path = filepath.Join(dir, filename)
+
+	if fi, err := os.Stat(dir); err == nil {
+		if !fi.Mode().IsDir() {
+			log.Printf("Invalid log directory '%s' mode\n", dir)
+			return DefaultLogPath
+		}
+	} else if errors.Is(err, os.ErrNotExist) {
+		err := os.MkdirAll(dir, 0755)
+
+		if err != nil {
+			log.Printf("Failed to create log directory '%s': %v\n", dir, err)
+			return DefaultLogPath
+		}
+	}
+
+	if fi, err := os.Stat(path); err == nil {
+		if !fi.Mode().IsRegular() {
+			log.Printf("Log file '%s' is not a regular file\n", path)
+			return DefaultLogPath
+		}
+	}
+
+	if !isWritable(path) {
+		log.Printf("Log file '%s' is not writable\n", path)
+		return DefaultLogPath
+	}
+
+	return path
```

Comment:
There aren't many checks here, to be honest.

If the file does not exist, perhaps even the directory does not, and if we can create, it makes sense to create instead of failing.

If the file is not regular (e.g. a device) it might still be writable, but we don't really want that.

The only check that we could remove is the relative path, if it is relative, I could simply return the default path.

If all checks pass it attempts to open the log file.

Reply:
I reworked this to be a bit cleaner and return an error instead of a valid file path.